h2o-algos/src/main/java/hex/ensemble/StackedEnsemble.java

```diff
@@ -143,108 +145,83 @@ private Frame prepareLevelOneFrame(String levelOneKey, Model[] baseModels, Frame
+    
+    protected Frame buildPredictionsForBaseModel(Model model, Frame frame) {
+      Key<Frame> predsKey = buildPredsKey(model, frame);
+      Frame preds = DKV.getGet(predsKey);
+      if (preds == null) {
+        preds =  model.score(frame, predsKey.toString());
+        Scope.untrack(preds.keysList());
+      }
+      if (_model._output._base_model_predictions_keys == null)
+        _model._output._base_model_predictions_keys = new NonBlockingHashSet<>();
+      
+      _model._output._base_model_predictions_keys.add(predsKey.toString());
+      //predictions are cleaned up by metalearner if necessary
+      return preds;
     }
 
+    protected abstract StackedEnsembleModel.StackingStrategy strategy();
+    
+    protected abstract Frame getActualTrainingFrame();
+    
+    protected abstract Frame getPredictionsForBaseModel(Model model, Frame actualsFrame, boolean isTrainingFrame);
+    
     private Key<Frame> buildPredsKey(Key model_key, long model_checksum, Key frame_key, long frame_checksum) {
       return Key.make("preds_" + model_checksum + "_on_" + frame_checksum);
     }
 
-    private Key<Frame> buildPredsKey(Model model, Frame frame) {
+    protected Key<Frame> buildPredsKey(Model model, Frame frame) {
       return frame == null || model == null ? null : buildPredsKey(model._key, model.checksum(), frame._key, frame.checksum());
     }
 
-    /**
-     * Prepare a "level one" frame for a given set of models and actuals.  Used for preparing validation frames
-     * for the metalearning step, and could also be used for bulk predictions for a StackedEnsemble.
-     */
-    private Frame prepareValidationLevelOneFrame(String levelOneKey, Key<Model>[] baseModelKeys, Frame actuals) {
-      List<Model> baseModels = new ArrayList<>();
-      List<Frame> baseModelPredictions = new ArrayList<>();
-
-      for (Key<Model> k : baseModelKeys) {
-        Model aModel = DKV.getGet(k);
-        if (null == aModel)
-          throw new H2OIllegalArgumentException("Failed to find base model: " + k);
-
-        Key<Frame> predsKey = buildPredsKey(aModel, actuals);
-        Frame aPred = aModel.score(actuals, predsKey.toString()); // TODO: cache predictions
-
-        baseModels.add(aModel);
-        if (!aModel._output.isMultinomialClassifier()) {
-          baseModelPredictions.add(aPred);
-        } else {
-          List<String> predColNames = new ArrayList<>(Arrays.asList(aPred.names()));
-          predColNames.remove("predict");
-          String[] multClassNames = predColNames.toArray(new String[0]);
-          baseModelPredictions.add(aPred.subframe(multClassNames));
-        }
-      }
-
-      Frame levelOne = prepareLevelOneFrame(levelOneKey, baseModels.toArray(new Model[0]), baseModelPredictions.toArray(new Frame[0]), actuals);
-
-      // remove baseModelPredictions frames and all the non-preds vecs from the DKV
-      for (Frame aPred : baseModelPredictions)
-        Frame.deleteTempFrameAndItsNonSharedVecs(aPred, levelOne);
```

Comment:
please note that those predictions have now a slightly longer lifetime: they're cleaned up by default only once the metalearner model is computed.